lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,239 @@
+
+        overall_bundle.add_metric('failed', [
+            name for name, bundle in res_bundles.items()
+            if bundle.result is Result.FAILED
+        ])
+        return overall_bundle
+
+class SchedTuneFreqItem(SchedTuneItemBase):
+    """
+    Runs a tiny RT rtapp task pinned to a big CPU at a given boost level and
+    checks the frequency selection was performed accordingly.
+    """
+
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "cpu_frequency"]
+    }, __qualname__)
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
+        rtapp_profile = {}
+        rtapp_profile['rta_stune'] = Periodic(
+            duty_cycle_pct = 1, # very small task, no impact on freq w/o boost
+            duration_s = 10,
+            period_ms = 16,
+            cpus = [cpu], # pin to big CPU, to focus on frequency selection
+            sched_policy = 'FIFO' # RT tasks have the boost holding feature so
+                                  # the frequency should be more stable, and we
+                                  # shouldn't go to max freq in Android
+        )
+        return rtapp_profile
+
+    def trace_window(self, trace):
+        """
+        Set the boundaries of the trace window to ``cpu_frequency`` events
+        before/after the task's start/end time
+        """
+        rta_start, rta_stop = super().trace_window(trace)
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        fdf = trace.df_events('cpu_frequency')
+        fdf = fdf[fdf.cpu == cpu]
+
+        # Find the frequency events before and after the task runs
+        freq_start = fdf[fdf.index < rta_start].index[-1]
+        freq_stop = fdf[fdf.index > rta_stop].index[0]
+
+        return (freq_start, freq_stop)
+
+    @requires_events('cpu_frequency')
+    def test_stune_frequency(self, freq_margin_pct=10) -> ResultBundle:
+        """
+        Test that frequency selection followed the boost
+
+        :param: freq_margin_pct: Allowed margin between estimated and measured
+            average frequencies
+        :type freq_margin_pct: int
+
+        Compute the expected frequency given the boost level and compare to the
+        real average frequency from the trace.
+        Check that the difference between expected and measured frequencies is
+        no larger than ``freq_margin_pct``.
+        """
+        kernel_version = self.plat_info['kernel']['version']
+        if kernel_version.parts[:2] < (4, 14):
+            self.get_logger().warning('This test requires the RT boost hold, but it may be disabled in {}'.format(kernel_version))
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        freqs = self.plat_info['freqs'][cpu]
+        max_freq = max(freqs)
+
+        # Estimate the target frequency, including sugov's margin, and round
+        # into a real OPP
+        boost = self.boost
+        target_freq = min(max_freq, max_freq * boost / 80)
```

Comment:
You sure about this formulation?

Is not overestimating a bit the target frequency, even before considering OPPs I mean?

Why not using the exact SPC formula instead?
Assuming a negligible utilization, something like:
```python
target_freq = 1.2 * max_freq * boost / 100 = max_freq * boost * 0.012
target_freq = min(target_freq, max_freq)
```
should give us the exact value with a 20% margin on top.

Reply:
Yeah but as per `map_util_freq()` what we apply is actually +25%:

 ```C
static inline unsigned long map_util_freq(unsigned long util,
                                        unsigned long freq, unsigned long cap)
{
        return (freq + (freq >> 2)) * util / cap;
}
```

So dividing by 0.8 is correct.